source/engine/graphics/components/SkeletalAnimator.cpp

```diff
@@ -1,45 +1,204 @@
-      m_onPlayStateChanged{},
-      m_initialState{anim::State{anim::Initial{std::move(animationUid)}}},
-      m_initialStateInitialized{}{}
+      m_immediateState{},
+      m_activeState{m_states.Insert(anim::State{anim::Root{std::move(animationUid)}})},
+      m_immediateStateSet{},
+      m_rootStateSet{}
+{
+}
 
 void SkeletalAnimator::Execute()
 {
-    if (ExecuteInitialState())
+    if (ExecuteRootState()) { return; }
+    if (ExecuteImmediateState()) { return; }
+    if (ExecuteExitState()) { return; }
+
+    for (auto childStateHandle : m_states.Get(m_activeState)->successors)
+    {
+        if (ExecuteChildEnterState(childStateHandle)) { return; }
+    }
+}
+
+auto SkeletalAnimator::AddState(anim::Loop loopProperties) -> uint32_t
+{
+    const auto addedStateHandle = m_states.Insert(anim::State{std::move(loopProperties)});
+    auto* addedState = m_states.GetLast();
+    m_states.Get(addedState->enterFrom)->AddSuccessor(addedStateHandle);
+    return addedStateHandle;
+}
+
+auto SkeletalAnimator::AddState(anim::PlayOnce playOnceProperties) -> uint32_t
+{
+    const auto addedStateHandle = m_states.Insert(anim::State{std::move(playOnceProperties)});
+    auto* addedState = m_states.GetLast();
+    m_states.Get(addedState->enterFrom)->AddSuccessor(addedStateHandle);
+    return addedStateHandle;
+}
+
+auto SkeletalAnimator::AddState(anim::Stop stopProperties) -> uint32_t
+{
+    auto state = anim::State{std::move(stopProperties)};
+    state.animUid = m_states.Get(m_activeState)->animUid;
+    const auto addedStateHandle = m_states.Insert(std::move(state));
+    auto* addedState = m_states.GetLast();
+    m_states.Get(addedState->enterFrom)->AddSuccessor(addedStateHandle);
+    return addedStateHandle;
+}
+
+void SkeletalAnimator::LoopImmediate(std::string animUid, std::function<bool ()> exitWhen, uint32_t exitTo)
+{
+    m_immediateState = std::make_unique<anim::State>(anim::Loop
+    {
+        .enterFrom = anim::NullState,
+        .enterWhen = anim::Never,
+        .animUid = std::move(animUid),
+        .exitWhen = std::move(exitWhen),
+        .exitTo = exitTo
+    });
+}
+
+void SkeletalAnimator::PlayOnceImmediate(std::string animUid, uint32_t exitTo)
+{
+    m_immediateState = std::make_unique<anim::State>(anim::PlayOnce
+    {
+        .enterFrom = anim::NullState,
+        .enterWhen = anim::Never,
+        .animUid = std::move(animUid),
+        .exitTo = exitTo
+    });
+}
+
+void SkeletalAnimator::StopImmediate(std::function<bool ()> exitWhen, uint32_t exitTo)
+{
+    m_immediateState = std::make_unique<anim::State>(anim::StopImmediate
+    {
+        .exitWhen = std::move(exitWhen),
+        .exitTo = exitTo
+    });
+}
+
+void SkeletalAnimator::CompleteFirstRun()
+{
+    if (m_immediateState && m_immediateState->action == anim::Action::PlayOnce)
+    {
+        *m_immediateState->firstRunComplete.get() = true;
+    }
+
+    if (m_states.Get(m_activeState) && m_states.Get(m_activeState)->action == anim::Action::PlayOnce)
```

Comment:
The compiler cannot see that `Get()` doesn't modify anything (without link-time optimization) and must leave all three calls in. I'd cache the result in a local variable.